### PR TITLE
[FIX] sale_margin: fix compute_purchase_price with use case timesheet_margin

### PR DESCRIPTION
**Pull request by yosa-odoo**
Steps to reproduce:
(sale_timesheet_margin, sale_stock_margin, sale_margin)
1. Sales -> configuration -> setting -> Activated the “Margin”.
2. Create a Product. ( Test Product)
- Define Sales Price, cost
- Product category select “Costing Method: Average Cost (AVCO)” and “Inventory Valuation: Automated”
3. Sales quotation
- select Product “Test Product”.
- Change the “Cost” value.
- Click on “Send by Email” (don’t click on the save button).
- Send an email.

Issue:
-> The cost has been reinitialised to its standard_price value.

Cause:
Everything's okay until you install `sale_timesheet_margin` in which case, whenever you send an email`you will compute the fields of the sale order and by extension the analytic_line which in turn triggers the `_compute_purchase_price` method from the sale_timesheet_margin
https://github.com/odoo/odoo/blob/e260a97392e698c88741be713a6b8cc2dc3c74a1/addons/sale_timesheet_margin/models/sale_order_line.py#L8-L9
And in
https://github.com/odoo/odoo/blob/e260a97392e698c88741be713a6b8cc2dc3c74a1/addons/sale_timesheet_margin/models/sale_order_line.py#L14

The method called in `sale_margin` has no case to keep the `purchase_price` if it has been set.

opw-2812360

### CODE REVIEW

**Review by agr-odoo — CHANGES_REQUESTED** [review]
Usual problem with stored-computed field
If we have already a purchase_price you may just skip the entire function
```
     def _compute_purchase_price(self):
         for line in self:
+            if line.purchase_price:
+                continue
```

Beside that I would reword the commit title in something more human readable like
`[FIX] sale_margin: do not overwrite user defined cost` or
`[FIX] sale_margin: keep user defined product_cost`

**Comment by yosa-odoo** [comment]
@agr-odoo thanks for the review. I applied the changes accordingly :)

**Review by agr-odoo — APPROVED** [review]
Looks ok to me
With this patch the system will not overwrite the purchase price set by user
@Feyensv

File: addons/sale_margin/models/sale_order.py [hunk 1]
```diff
@@ -19,6 +19,8 @@ class SaleOrderLine(models.Model):
     @api.depends('product_id', 'company_id', 'currency_id', 'product_uom')
     def _compute_purchase_price(self):
         for line in self:
+            if line.purchase_price:
+                continue
```
Comment (Feyensv) [inline addons/sale_margin/models/sale_order.py:23]: Nope, we want the recomputation when the product/company/currency/uom is modified...
If there is a specific timesheet case, it should avoid calling super for the line where the recomputation shouldn't happen, but not disable the computation.
With your change, the compute dependencies simply become useless :confused: (except for the case when the computed price was 0).

Comment (yosa-odoo) [reply addons/sale_margin/models/sale_order.py]: I'll work on a better solution that does not skip the computation but keep the purchase_price is set.

I will also create a test.

Thanks for the review!

**Comment by yosa-odoo** [comment]
@Feyensv I added a test and slightly change the fix so it goes through computation.
It passes the test with the fix - It does not pass the test without the fix

File: addons/sale_margin/models/sale_order.py [hunk 2]
```diff
@@ -24,6 +24,8 @@ def _compute_purchase_price(self):
                 continue
             line = line.with_company(line.company_id)
             product_cost = line.product_id.standard_price
+            if line.purchase_price:
+                product_cost = line.purchase_price
```
Comment (Feyensv) [inline addons/sale_margin/models/sale_order.py:28]: > we want the recomputation when the product/currency/uom is modified.

AFAIK, if I change the product now, the cost won't be recomputed...

Honestly, after more thoughts, I don't know if there is a way to fix this problem (we would have to not trigger the recomputation for non timesheet lines if the computation trigger is the timesheet dependencies), but if there is, it must be done in sale_timesheet_margin, the base behavior is the expected one, you shouldn't have to modify it...

File: addons/sale_timesheet_margin/tests/test_sale_order.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+
+class TestSaleTimesheetMargin():
+
+    @classmethod
+    def setUpClass(cls):
+        pass
```
Comment (Feyensv) [inline addons/sale_timesheet_margin/tests/test_sale_order.py:9]: I believe your test isn't triggered (otherwise it would fail, as it doesn't follow the structure of odoo tests).
Did you at least take a look at other tests to create the base of your test class ?

**Review by Feyensv — CHANGES_REQUESTED** [review]

File: addons/sale_timesheet_margin/tests/test_sale_order.py (hunk 3 above)
Comment (yosa-odoo) [reply addons/sale_timesheet_margin/tests/test_sale_order.py]: I did a bit of cleaning before committing to be minimalist. But it was wrong...

**Comment by yosa-odoo** [comment]
Hello @Feyensv ,
Since it seems that we are facing a technical limitation. Wouldn't it make sense to make the field purchase_price "Read-Only" ?

**Comment by Feyensv** [comment]
Only when `sale_timesheet_margin` is installed then, to avoid restricting the customers when not necessary.

Did you try investigating the exact place where the recomputation is triggered and to remove the lines that shouldn't be recomputed from the computation queue ? (with env.remove_to_compute)

**Comment by yosa-odoo** [comment]
After some back and forth, the change has been considered not desirable.

```
In FIFO and AVCO, the cost is always recomputed and until the last step of the flow. It is in order to have the most accurate cost.

In 14.0, if you send an email, the modified cost is maintained but when you confirm the SO, the cost will be recomputed.
Side note: In some countries, it is even legally binding.
```
